### DISCUSSION TITLE
How to calibrate weights of models after apply Accelerating a Hugging Face Llama 3 models with Transformer Engine

### DISCUSSION THREAD

**Discussion by xuanvinh1997** [NVIDIA/TransformerEngine#1748] [Q&A]
After reading this tutorial, [Accelerating a Hugging Face Llama 2 and Llama 3 models with Transformer Engine](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/te_llama/tutorial_accelerate_hf_llama_with_te.html), I now understand how to convert HF/Torch layers to TE layers and fuse them. However, the model's output is no longer correct.

My question is: how can I calibrate it without re-training the pretrained model?

Many thanks!

**Comment by Advait251206** [comment]
This is a very common confusion when first moving a Hugging Face Llama model to Transformer Engine.

### Short answer
You generally **do not “calibrate weights” after converting a pretrained HF Llama model to Transformer Engine (TE)**.

If outputs become incorrect, it is almost always due to **missing precision handling or missing FP8 calibration (scaling statistics), not weight calibration** in the classical sense.

---

## Why your model output breaks

When you “convert + fuse” HF Llama → TE, you are usually doing two things:

1. Replacing `torch.nn.Linear` with `te.Linear`
2. Enabling FP8 autocast / FP8 GEMMs

The key issue is:

> FP8 does NOT work without collecting activation statistics (amax history) to compute scaling factors.

If this step is missing or incorrect, you get:
- garbage outputs
- NaNs
- severe accuracy degradation

---

## What “calibration” actually means in Transformer Engine

In TE, calibration usually refers to:

### ✅ FP8 amax (activation max) collection
Transformer Engine builds scaling factors using:
- activation `amax`
- weight `amax`
- history over multiple steps (delayed scaling)

This is **not weight fine-tuning**
It is **numerical range calibration for FP8 scaling**

---

## Correct ways to fix / “calibrate” without training

### 1. Run a calibration phase (recommended)

Before trusting outputs, run several forward passes:

```python
with te.fp8_autocast(enabled=True, fp8_recipe=fp8_recipe):
    for batch in calibration_dataloader:
        _ = model(batch)
```

This allows TE to:
- collect amax statistics
- stabilize scaling factors
- warm up FP8 metadata

👉 This is REQUIRED for stable FP8 inference/training.

---

### 2. Use BF16 first to validate correctness

Before FP8:

```python
with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
    outputs = model(inputs)
```

If BF16 output is already incorrect:
- conversion to TE is broken (not FP8 issue)

---

### 3. Ensure delayed scaling is enabled (critical)

Use:

```python
from transformer_engine.pytorch import recipe

fp8_recipe = recipe.DelayedScaling(
    fp8_format=recipe.Format.E4M3
)
```

Why this matters:
- avoids per-step instability
- uses historical amax instead of instant noisy values

---

### 4. Make sure all Linear layers are actually TE layers

A very common failure:

> Only part of the model is converted → mixed execution paths → wrong outputs

Check:

```python
import transformer_engine.pytorch as te

for name, module in model.named_modules():
    if isinstance(module, torch.nn.Linear):
        print("Unconverted layer:", name)
```

If any remain → FP8 path is incomplete.

---

### 5. Disable FP8 to verify conversion correctness

If TE conversion is correct, this should match HF model:

```python
with te.fp8_autocast(enabled=False):
    outputs = model(inputs)
```

If this still differs:
- conversion / weight mapping is wrong
- NOT an FP8 calibration issue

---

## Important clarification

### ❌ There is NO “post-hoc weight calibration”
Unlike quantization (like INT8 PTQ), Transformer Engine FP8:

- does NOT require weight retraining
- does NOT require dataset-based calibration of weights
- does NOT modify model semantics

It only calibrates:
> dynamic scaling factors (amax-based FP8 scaling)

---

## If outputs are still wrong after calibration

Then likely causes are:

- Missing layer conversions (some Linear layers still FP16/BF16)
- Incorrect attention / normalization fusion
- Wrong placement of `fp8_autocast` scope
- Using FP8 on unsupported ops (e.g., embeddings, softmax)
- Mixing HF generate() internals with partial FP8 scope

---

## Practical recommendation

For LLaMA 3 specifically:

1. First ensure BF16 correctness after conversion
2. Then run 50–200 warmup forward passes
3. Then enable FP8 inference
4. Only then benchmark outputs

---

## Summary

- There is **no weight calibration step**
- What you need is **FP8 amax/statistics calibration**
- This happens automatically during warmup forward passes
- Most “wrong output” issues come from **incorrect layer conversion or incomplete FP8 coverage**, not missing calibration

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
